src/server/pkg/storage/fileset/storage.go

```diff
@@ -178,52 +198,44 @@ func (s *Storage) CompactSpec(ctx context.Context, fileSet string, compactedFile
 	var level int
 	// Handle first commit being compacted.
 	if len(compactedFileSet) == 0 {
-		for size > s.levelZeroSize*int64(math.Pow(float64(s.levelSizeBase), float64(level))) {
+		for size > s.levelSize(level) {
 			level++
 		}
-		spec.Output = path.Join(fileSet, Compacted, strconv.Itoa(level))
+		spec.Output = path.Join(fileSet, Compacted, levelName(level))
 		return spec, nil
 	}
-	// Handle commits with a parent commit.
-	compactedFileSet[0] = applyPrefix(compactedFileSet[0])
-	if err := s.objC.Walk(ctx, path.Join(compactedFileSet[0], Compacted), func(name string) error {
-		nextLevel, err := strconv.Atoi(path.Base(name))
+	// While we can't fit it all in the current level
+	for {
+		levelPath := path.Join(compactedFileSet[0], Compacted, levelName(level))
+		idx, err := index.GetTopLevelIndex(ctx, s.objC, levelPath)
+		if err != nil && !s.objC.IsNotExist(err) {
+			return nil, err
+		}
+		// The level does exist. Add it's size to the total, and mark it for compaction.
+		if err == nil {
+			spec.Input = append(spec.Input, levelPath)
+			size += idx.SizeBytes
+		}
+		if size <= s.levelSize(level) {
+			break
+		}
+		level++
+	}
+	// Now we know the output level
+	spec.Output = path.Join(fileSet, Compacted, levelName(level))
+	// Copy the other levels that may exist
+	if err := s.objC.Walk(ctx, path.Join(compactedFileSet[0], Compacted), func(src string) error {
+		lName := path.Base(src)
+		l, err := parseLevel(lName)
 		if err != nil {
 			return err
 		}
-		// Handle levels that are non-empty.
-		if nextLevel == level {
-			idx, err := index.GetTopLevelIndex(ctx, s.objC, name)
-			if err != nil {
+		dst := path.Join(fileSet, Compacted, levelName(l))
```

Comment:
This should probably be setup in the below if block (it would be unused otherwise).